[Android][Sync]  Add advanced option in Sync page

jagadeshjai opened the pull request:
- Update the illustrate in Sync first page
- Add `Advanced options` button at the bottom and its dialog which handled input logics of self host sync URL.
- Add read only self host sync url at the top of devices listing screen.

<!-- Add brave-browser issue below that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/42570

# OutCome

https://github.com/user-attachments/assets/14c888a6-b03e-47bb-86e2-b31dd6f2a3ac



| Advanced option dialog | After entering self host sync URL | Non HTTPS Error msg |  Read Only self host URL |
|--------|--------|--------|--------|
| ![IMG_20241220_235331](https://github.com/user-attachments/assets/75eae7f8-3be2-47b6-ab45-4f1e4100a267) | ![IMG_20241220_235314](https://github.com/user-attachments/assets/bee74b44-d7f3-42d6-8077-eb5c1be64605)  | ![https_error_msg](https://github.com/user-attachments/assets/5f526338-0882-4e1f-9f41-cc365c795c07) | ![read_only_sync_url](https://github.com/user-attachments/assets/9af56638-c2ee-4e41-a54d-ed4284f28dd1) |




<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

### Fresh install / First time

1.  Navigate to Settings > Sync and make sure sync is turned off.
2. Click the `Advanced options` at the bottom and a dialog will be shown.
3.  The default sync URL should be displayed as a placeholder in the `Self-host sync URL` text box.
4. Enter a self host sync URL, which should an HTTPS URL and Click `Save and Relaunch`.
5.  Now navigate to  `brave://sync-internals`  and check whether the given self host URL shown under `Version Info > Server URL` . 
6. Once again navigate to  `Sync` page and start the sync using any one option(`Scan or enter sync code` or `Start a new sync chain`)
7. And Make sure `Self-host sync URL` is visible at the top of Device listing page once Sync is started.

### Error Handling

1. An error message saying `Invalid URL` should appear when an invalid URL is entered.
2. An error should also appear for non-HTTPS URLs.

AlexeyBarabash commented on android/java/org/chromium/chrome/browser/settings/BraveSyncScreensPreference.java:
```diff
@@ -1233,17 +1266,130 @@ public void onClick(DialogInterface dialog, int button) {
+        }
+
+        // Check if the URL starts with a valid scheme (http:// or https://)
+        if (!url.matches("^[a-zA-Z][a-zA-Z0-9+.-]*://.*")) {
```
please use [GURL.getScheme](https://source.chromium.org/chromium/chromium/src/+/main:url/android/java/src/org/chromium/url/GURL.java;l=186?q=GURL.java&ss=chromium%2Fchromium%2Fsrc)

jagadeshjai replied:
Sure.

jagadeshjai commented:
> Thanks, @jagadeshjai this is a great work.
> 
> There are several things can be polished to improve the PR.
> 

Thanks for the reviews @AlexeyBarabash ! I'll look into this.

AlexeyBarabash commented on components/sync/service/brave_sync_service_impl.h:
```diff
@@ -73,6 +73,9 @@ class BraveSyncServiceImpl : public SyncServiceImpl {
   void StopAndClearWithShutdownReason();
   void StopAndClearWithResetLocalDataReason();
 
+  std::string GetCustomSyncServiceURL();
```
It looks like we can move both these methods into `BraveSyncServiceImpl` class.
So we will have all the methods right in the same place  
```
  std::string GetCustomSyncServiceURL();
  bool SetCustomSyncServiceURL(std::string custom_sync_url);

  GURL GetBraveSyncServiceURL();
  GURL GetBraveDefaultSyncServiceURL();
  ```
It will be possible to add a test case for these methods.
  
It may will require make `BraveGetSyncServiceURL` the member of `SyncServiceImpl` `chromium_src/components/sync/service/sync_service_impl.cc`.

jagadeshjai commented:
@AlexeyBarabash Added the Test plan in PR description and fixed the mentioned UI issues, please take a look.

>     6. I saw you put the raw values into `android/java/res/values-night/brave_colors.xml`/`.../values-night/brave_colors.xml` values directly - there are no values corresponding to Figma colors at `android/java/res/values-night/brave_colors.xml`.
>        At this case we add the values from `brave-browser/src/brave/node_modules/@brave/leo/tokens/android/values/colors.xml`
>        I need to figure out with our team how to make it better.


As of now, I've added entries for missing colors in `values(-night)/brave_colors.xml` by referring the colors in `node_modules/@brave/leo/tokens/android/values/colors.xml`. o avoid confusion with existing names, I've prefixed them with `leo_color_` , Kindly let me know if you have any idea on this.

AlexeyBarabash reviewed (approved):
lgtm, great work, thanks @jagadeshjai

jagadeshjai commented on android/java/res/layout/brave_sync_done.xml:
```diff
@@ -40,18 +40,43 @@
+            android:paddingHorizontal="16dp"
+            android:visibility="gone">
+            <TextView
+                android:layout_width="328dp"
```
No, We can use match_parent, wrap_content.

deeppandya reviewed (approved):
LGTM. Amazing work @jagadeshjai. Thank you for the changes.

bridiver commented on browser/android/brave_sync_worker.cc:
```diff
@@ -232,6 +234,51 @@ void BraveSyncWorker::OnStateChanged(syncer::SyncService* service) {
   }
 }
 
+base::android::ScopedJavaLocalRef<jstring> BraveSyncWorker::GetSyncServiceURL(
```
this is not needed, prefs can be accessed directly in java

bsclifton commented:
Closing in favor of https://github.com/brave/brave-core/pull/28463
